designdocs/permissions.md

```diff
@@ -0,0 +1,87 @@
+# proposal for permissions request
+
+
+### Philosophy
+The user-agent will always be closer to the user than a spec. The user agent has greater access to current conditions than we (the spec writers) ever will. As a philosophy we should give the user agent *as much freedom as possible to “do the right thing”* under whatever circumstances are at the time a web application is started.
+
+### scope
+
+Currently we are having difficulties finding a solution that all parties can agree to. This is about the permissions for various properties in WebXR.  This proposal addresses the following concerns
+
+* what permissions are available to be asked for
+* how granular are these permissions
+* are the permissions granted at page load / application start
+* or are the permissions granted at the time of use.
+
+### stating concerns
+
+Concerns about granting at page load time are
+* users may be requested to approve multiple permissions one after another, creating click fatigue.
+* if all apps request permissions at start, the user has no information to judge whether or not they should grant this, because they haven’t entered the experience yet.
+
+Concerns about granting requests once already inside of immersive mode are:
+* access requests when inside immersive mode should be done with some sort of a secure method, such as a dedicated hardware button, so that applications cannot spoof such permissions.
+* some devices or scenarios may not be able to provide such a secure method, thus doing the request before entering immersive mode is preferable.
+
+Concerns about both systems include:
+* the user agents know more about the user, the current scenario, and can make better decisions *at the time of use* than a spec can. Neither method give the user agent freedom to adapt.
+* there may be new attacks devised or new flaws discovered after we have shipped, but user agents will not be able to address these because they are constrained by the spec.
+
+There are most likely more concerns than just what I have listed above.
+
+### Solution
+To address these I propose the following: *do both*.  Let the developers have a single code path where they /request permissions at application start/ *and* when /the permission is actually needed/. The first request is considered the upper bounds of possible permissions. The user agent can decide whether to actually show user request dialogs at app time or when the permission is actually used, or some other scheme that the user agent deems to be better (such as warding off new attacks in the future).
```

Comment:
Thanks @joshmarinacci - during our previous conversation you'd suggested a more granular set of permission requests which I took to mean something like:

_Required_ - needed for the session to run at all.
_Optional_ - not blocking for the session to run, but desirable.
_Deferrable_ - not required at session creation, but might be required depending upon what the user does

note: I made up those names because I couldn't remember exactly what you said, but some of this is covered in #424 

The intent as I understood it was to give guidance to the user agent on the timing of when user consent events would occur, and whether or not a session should be created if the user said 'no' to certain features.

Do you want to add that detail here? Can a user agent make the decisions outlined in this paragraph without this type of information?

Reply:
> note: I made up those names

About terminology, have you considered using a term other than "permissions" since this is somewhat different from existing web permissions? Please ignore if you've discussed this already. Elsewhere we've used terms such as "consent" to distinguish it, especially if this is likely to be a per-session prompt instead of permanently granting access to a site.

Reply:
Oh interesting... that's not the interpretation I took... I thought by granular we meant something like "spatial-tracking", "real-world-understanding", "camera-access", "eye tracking", etc.  The big question there is what are the buckets we should use?  That said, if we do take the modularization approach, the good news is that we'd only need to define/ bikeshed the one that would cover XRPose and XRViewerPose data for now, yeah?  Then, as we add other modules we can define the appropriate enum values.

The other interesting thing about this approach is that we have the opportunity to possibly integrate it with https://www.w3.org/TR/permissions/#enumdef-permissionname so that developers can request non-XR things up front as well!  That would address the concerns about developers wanting to prompt for permissions for things like Microphone mid-immersive session :) 

That said... I've been consistently getting the impression that folks from Google have specific concerns about the Permissions API given the number of times y'all have mentioned wanting to avoid the word "permissions" in favor of "user consent".  Is there some background or context that I'm missing that makes it contentious?

Reply:
Actually, rereading @johnpallet's privacy doc, I'm realizing it's probably more like "spatial-tracking" and "spatial-tracking-unbounded" that would be needed for the core WebXR spec?

Reply:
> That said... I've been consistently getting the impression that folks from Google have specific concerns about the Permissions API given the number of times y'all have mentioned wanting to avoid the word "permissions" in favor of "user consent". Is there some background or context that I'm missing that makes it contentious?

Nell, the privacy explainer talks about some of the concerns and distinctions in the [permissions](https://github.com/immersive-web/privacy-and-security/blob/master/EXPLAINER.md#permissions) section,

I wouldn't necessarily call it contentious, but I think it's helpful to make a distinction between the more abstract concept of "how do we get informed user consent for using these features" from a specific implementation method as in "let's use the browser's existing Permissions API support for this". While the permissions API spec is fairly general, the practical implementations in web browsers tend to be based around persistently saved per-site choices, and as the privacy doc explains this isn't necessarily a good fit to some of the AR/VR use cases. I think it's better to avoid the word "permissions" at least during the design phase to avoid conflating these two interpretations.

Personally I'm in favor of using the Permissions API where it makes sense, but AFAIK it hasn't commonly been used for use cases such as per-session temporary permissions. That seems compatible with the spec by slightly bending the "new information about the user’s intent" rule as applying when a session ends, but I think that's not what people would typically have in mind when talking about web permissions, and it may need additional implementation work in browsers to generalize internal permission handling to make this work.

Reply:
To be honest, when I originally read that document when it was completed, the distinction between "permissions" and "user consent" didn't jump out as being meaningful.  And reading it again, I'm still not seeing it?  Definitely not trying to be obstinate; just trying to get on the same page.  When "permissions" are being referred to... it's what?  The UI each browser has chosen to implement up to this point in time?  I think my confusion is coming from my understanding that the permissions spec defines an algorithm for gathering consent but doesn't appear to be prescriptive about how the UA actually visualizes the request?  Or did I misread it?

Either way though....given that we are intending to hit "VR complete" in 17 days, it seems like time to make a call on whether or not we have any form of dependency on the Permissions spec.  We're sort of running out of runway for @toji and I to write, get consensus, and merge explainer/spec text to cover the privacy, security, and user consent problem space.  So if a "design doc" isn't where folks would like to have this discussion, is there somewhere else preferable?  I've been avoiding putting together a PR against the explainer since we aren't in agreement on the core principles of the approach, but if that would work better for folks I can do so?

Reply:
I wasn't await of the Permissions spec until now. Interesting. Would it's query() function violate our desire to avoid fingerprinting? As long as query() is called after the initial session request would it still be okay?

Reply:
@ddorwin Thanks for the pointer to the MediaKeysRequirement stuff! time to go fall down the rabbit hole of spelunking through a new spec!

Reply:
@joshmarinacci That's basically the question I'm asking.  If it's ok for Window, would it be ok if had a XRSession.Permissions.query()?

Reply:
In the privacy design doc the term 'permission' was intentionally avoided in part to avoid inheriting the algorithms and requirements of the [Permission API](https://www.w3.org/TR/permissions/), and also because 'permissions' are an established concept in browsers. It's still not clear to me whether using the term 'permissions' leaves sufficient room for user consent flows such as the screenshot from @blairmacintyre (a dialog with multiple toggles), or user consent that is valid only for the duration of a session (vs. causing a more persistent state change). For those reasons the privacy design doc instead uses the term 'user consent' to describe the requirement.

@joshmarinacci my original question got a bit lost in this thread. Did you intend to have 'required', 'optional' and 'maybe later' as three different options when requesting features?